### ISSUE TITLE
Add support for Connect RPC protocol

### ISSUE THREAD

**Issue by akshayjshah** [envoyproxy/envoy#22679]
*Title*: *Add support for the Connect RPC protocol, on par with gRPC and gRPC-Web*

*Description*:
Add two features to support the [Connect RPC protocol](https://connect.build/docs/protocol):

1. Export metrics for proxied Connect traffic, as Envoy does today for gRPC.
2. Add a filter to translate from the Connect protocol to gRPC, as Envoy does today to expose gRPC backends to gRPC-Web clients.

The Connect protocol and the associated RPC libraries are similar to gRPC: they use HTTP as a transport, support unary and streaming RPCs, and are encoding agnostic but usually paired with Protobuf. Unlike gRPC, the Connect protocol works with HTTP/1.1,<sup>&dagger;</sup> doesn't use HTTP trailers, and keeps the wire representation of unary RPCs [dead simple](https://connect.build/docs/protocol#unary-examples). Because it demands very little of the underlying HTTP transport, Connect is easy to adopt: clients can be as simple as a cURL one-liner, and popular frameworks like Rails, Flask, and Laravel can serve Connect traffic.

Envoy's excellent metrics are, by themselves, a significant incentive for organizations to adopt gRPC. It'd be wonderful to bring the same consistent observability to Connect traffic.

Of course, many organizations already have large gRPC deployments. A translating filter would let web browsers call these backends using a protocol that's more developer-friendly than gRPC-Web. We designed Connect with translating proxies in mind, so converting to and from gRPC is simple, reasonably efficient, and semantically lossless.

Thanks for considering this enhancement!

*Relevant Links*:
- The [Connect website](https://connect.build), especially the [project overview](https://connect.build/docs/introduction)
- The [Connect protocol specification](https://connect.build/docs/protocol)
- The [Go implementation](https://github.com/bufbuild/connect-go)
- The [TypeScript implementation](https://github.com/bufbuild/connect-web) (for browsers)
- Implementations for Node.js, Kotlin, and Swift are in progress.

<hr>
<sup>&dagger;</sup> Naturally, bidirectional streaming still requires HTTP/2. Unary, server streaming, and client streaming RPCs work over HTTP/1.1 or HTTP/2.

**Comment by wbpcode** [comment]
cc @alyssawilk

**Comment by mattklein123** [comment]
I'm in favor of supporting this, pending finding someone to implement it. It should be very similar to the grpc-web filter. One thing that I wonder is whether the grpc-web filter can just be modified to also support connect traffic which might make overall maintenance less costly.

**Comment by akshayjshah** [comment]
Glad to hear it :tada: If there's agreement that these additional features are welcome, we can help find bandwidth to implement:

* Someone at @bufbuild could put up a PR. Our skill ranges from "saw some C++ decades ago" to "wrote a bit of production C++ 3y ago," but nobody on our team is fearlessly shipping GFE diffs 🤣 That said, we can make sure that the Connect protocol gets implemented correctly.
* If it would help the Envoy community, we could also offer a short-term contract to someone in the community. Perhaps someone's between full-time jobs or wants to start putting together an Envoy-focused contracting business?

Basically, we're happy to partner up on this.

**Comment by akshayjshah** [comment]
@jchadwick-buf has a draft PR open (https://github.com/envoyproxy/envoy/pull/24836) to add two filters to Envoy: one to collect stats for Connect traffic, and one to bridge Connect to gRPC.

In the few months since we opened this issue, adoption of Connect has grown a fair bit. Privately, a number of users have said that they'd like these features in Envoy.

* Envoy's gRPC stats are excellent, so organizations often grow to depend on them. Orgs moving from gRPC to Connect naturally want the same stats, collected in the same place.
* Quite a few mobile and web users of `connect-{web,kotlin,swift}` were reluctant to adopt the grpc-web protocol, but are happy with Connect. They'd like to drop their cumbersome grpc-gateway deployments in favor of a simple Connect-to-gRPC bridge within Envoy, which they'll deploy as an API gateway.
* Backend microservice teams have begun migrating to the Connect protocol, but they'd like a shim to let their existing gRPC-only servers speak the new protocol. Many of those users already run Envoy as part of their service mesh's data plane, so the bridge filter works well for them.

As far as I know, only [PlanetScale](https://planetscale.com/blog/faster-mysql-with-http3) is currently discussing their use of Connect in public. I'll poll our Slack and see if any other users are able to publicly +1 this comment.

* The bridge filter has been a consistent request from orgs migrating from gRPC to Connect. Mobile and web teams want it because they like the Connect protocol, but need to interoperate with a large gRPC-only backend. (Switching to the c
